signac/contrib/_index.py

```diff
@@ -0,0 +1,312 @@
+    any of the same ways as a :class:`dict`, like ``_Index(**kwargs)``,
+    ``_Index(mapping, **kwargs)``, or ``_Index(iterable, **kwargs)``.
+
+    """
+
+    def build_index(self, key):
+        """Build index for a given key.
+
+        This is a highly performance critical code path.
+
+        Parameters
+        ----------
+        key : str
+            The key on which the index is built.
+
+        Returns
+        -------
+        :class:`~_TypedSetDefaultDict`
+            Index for key.
+
+        Raises
+        ------
+        :class:`~signac.errors.InvalidKeyError`
+            The document contains invalid keys.
+
+        """
+        logger.debug(f"Building index for key '{key}'...")
+        nodes = key.split(".")
+        index = _TypedSetDefaultDict()
+
+        for _id, doc in self.items():
+            try:
+                v = doc
+                # Recursively access nested values from dotted keys.
+                for n in nodes:
+                    v = v[n]
+            except (KeyError, TypeError):
+                pass
+            else:
+                if type(v) is list:  # Avoid isinstance for performance.
+                    index[_to_hashable(v)].add(_id)
+                else:
+                    if type(v) is dict:  # Avoid isinstance for performance.
+                        v = _DictPlaceholder
+                    index[v].add(_id)
+
+            # If the original key has dots and is present, raise an error.
+            if len(nodes) > 1 and key in doc:
+                raise InvalidKeyError(
+                    "Keys with dots ('.') are invalid.\n\n"
+                    "See https://signac.io/document-wide-migration/ "
+                    "for a recipe on how to replace dots in existing keys."
+                )
+        logger.debug(f"Built index for key '{key}'.")
+        return index
+
+    def _find_expression(self, key, value):
+        """Find document for key value pair.
+
+        Parameters
+        ----------
+        key : str
+            The key for expression.
+        value :
+            The value for expression.
+
+        Returns
+        -------
+        set
+            The document for key value pair.
+
+        Raises
+        ------
+        KeyError
+            An invalid operator was given.
+        ValueError
+            The value is not bool for '$exists' operator or not a
+            supported type for '$type' operator.
+
+        """
+        logger.debug(f"Find documents for expression '{key}: {value}'.")
+        if "$" in key:
+            if key.count("$") > 1:
+                raise KeyError(f"Invalid operator expression '{key}'.")
+            nodes = key.split(".")
+            op = nodes[-1]
+            if not op.startswith("$"):
+                raise KeyError(f"Invalid operator placement '{key}'.")
+            key = ".".join(nodes[:-1])
+            if op in _INDEX_OPERATORS:
+                index = self.build_index(key)
```

Comment:
To be clear, this is the kind of operation that I was hoping to avoid when I suggested changing the build_index API. We rejoin the string just to split it again immediately inside build_index. That sequence of operations isn't exactly super expensive, but it's also not free, especially given that we see the impact of microoptimizations on the level `isinstance` vs `type` checks:
```
In [20]: %timeit '.'.join('a.b.c'.split('.')[1:]).split('.')
263 ns ± 22.2 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each
```

Reply:
This split occurs `O(number of keys in a filter)` times. Most of the other microoptimizations happen `O(jobs)` times, which are the cases I care about most. I am going to just resolve this and say it's good enough for now.